kong/runloop/plugin_servers/pb_rpc.lua

```diff
@@ -0,0 +1,426 @@
+        list_v = structpb_list(v)
+      else
+        struct_v = structpb_struct(v)
+      end
+    end
+
+    return {
+      null_value = t == "nil" or nil,
+      bool_value = bool_v,
+      number_value = t == "number" and v or nil,
+      string_value = t == "string" and v or nil,
+      list_value = list_v,
+      struct_value = struct_v,
+    }
+  end
+
+  function structpb_list(l)
+    local out = {}
+    for i, v in ipairs(l) do
+      out[i] = structpb_value(v)
+    end
+    return { values = out }
+  end
+
+  function structpb_struct(d)
+    local out = {}
+    for k, v in pairs(d) do
+      out[k] = structpb_value(v)
+    end
+    return { fields = out }
+  end
+
+  local function wrap_val(v) return { v = v } end
+
+  pb_wrap = {
+    [".google.protobuf.Value"] = structpb_value,
+    [".google.protobuf.ListValue"] = structpb_list,
+    [".google.protobuf.Struct"] = structpb_struct,
+
+    [".kong_plugin_protocol.Bool"] = wrap_val,
+    [".kong_plugin_protocol.Number"] = wrap_val,
+    [".kong_plugin_protocol.Int"] = wrap_val,
+    [".kong_plugin_protocol.String"] = wrap_val,
+    --[".kong_plugin_protocol.MemoryStats"] = - function(v)
+    --  return {
+    --    lua_shared_dicts = {
+    --
+    --    }
+    --  }
+    --end
+  }
+end
+
+
+local function index_table(table, field)
+  if table[field] then
+    return table[field]
+  end
+
+  local res = table
+  for segment, e in ngx.re.gmatch(field, "\\w+", "o") do
+    if res[segment[0]] then
+      res = res[segment[0]]
+    else
+      return nil
+    end
+  end
+  return res
+end
+
+local function load_service()
+  local exposed_api = {
+    kong = kong,
+
+    ["kong.log.serialize"] = function()
+      local saved = Rpc.save_for_later[coroutine.running()]
+      return cjson_encode(saved and saved.serialize_data or kong.log.serialize())
+    end,
+
+    ["kong.nginx.get_var"] = function(v)
+      return ngx.var[v]
+    end,
+
+    ["kong.nginx.get_tls1_version_str"] = ngx_ssl.get_tls1_version_str,
+
+    ["kong.nginx.get_ctx"] = function(k)
+      local saved = Rpc.save_for_later[coroutine.running()]
+      local ngx_ctx = saved and saved.ngx_ctx or ngx.ctx
+      return ngx_ctx[k]
+    end,
+
+    ["kong.nginx.set_ctx"] = function(k, v)
+      local saved = Rpc.save_for_later[coroutine.running()]
+      local ngx_ctx = saved and saved.ngx_ctx or ngx.ctx
+      ngx_ctx[k] = v
+    end,
+
+    ["kong.ctx.shared.get"] = function(k)
+      local saved = Rpc.save_for_later[coroutine.running()]
+      local ctx_shared = saved and saved.ctx_shared or kong.ctx.shared
+      return ctx_shared[k]
+    end,
+
+    ["kong.ctx.shared.set"] = function(k, v)
+      local saved = Rpc.save_for_later[coroutine.running()]
+      local ctx_shared = saved and saved.ctx_shared or kong.ctx.shared
+      ctx_shared[k] = v
+    end,
+
+    ["kong.nginx.req_start_time"] = ngx.req.start_time,
+  }
+
+  local p = protoc.new()
+  --p:loadfile("kong/pluginsocket.proto")
+
+  p:addpath("/usr/include")
```

Comment:
hardcode this path feels slightly weird, when/who will copy the proto file into this path?

Reply:
right.  there's where the "standard" google structs are (on my system)  should we include a copy of that? add to the deploy scripts?

Reply:
> right. there's where the "standard" google structs are (on my system) should we include a copy of that? add to the deploy scripts?

I think we have at least two options here:
- Marking them as a package dependency (if a package for it exists in the systems we support)
- Bundling them into our usual `/usr/local/kong/` or `/usr/local/kong/lib` paths

Reply:
Does it make sense to include them in kong's src tree? The import files are not so large in size but i'm not sure if that's a common practice, since unlinke other implementations lua-protobuf doesn't bundle any google proto by default.

Reply:
just checked, for ubuntu it's `libprotobuf-dev`

Reply:
We have talked about this online and we'll call this a "known limitation" of the pb plugins server, will try to improve it after the Beta.